File: journal_analysis.py
#!/usr/bin/env python3
from collections import Counter
import os
import re
import random
from string import punctuation
import sys
import time

start = time.time()
entries = []
wordcount = Counter()
string_file = []
file_dict = {}
r = re.compile(r'[\s]+'.format(re.escape(punctuation)))

for root, dirs, files in os.walk("/mnt/c/Eric/Journal"):
    for file in files:
        file_words = []
        if file.endswith(".txt"):
            if file[0].isdigit():
                entries.append(file)
            with open(os.path.join(root, file), "r", encoding='UTF-8') as f:
                # Store an entire file into a string
                string_file = f.read()
                file_dict[file] = string_file
            with open(os.path.join(root, file), "r", encoding='UTF-8') as f:
                # Get all the words in a file
                file_words = f.read().split()
                # find file that specific words appear in
                # if '' in file_words:
                #    print(file)
        # Regex to split on punctation
        split_words = [r.split(s) for s in file_words]
        # Flatten list
        flat_words = [item for subitem in split_words for item in subitem if item]
        wordcount += Counter(flat_words)

print('Took: ', time.time() - start, ' seconds')

# Get a random journal entry
length_of_passage = 500
passage_date, random_passage = random.sample(file_dict.items(), 1)[0]
randomNum = random.randint(0, len(random_passage) - 100 + 1)
print('Date: ', passage_date)
print(random_passage[randomNum : randomNum + length_of_passage])

print('Number of entries: ', len(entries))
print('Total words: ', sum(wordcount.values()))
print('Unique words: ', len(list(wordcount)))

most_comm = wordcount.most_common(500)
# Find most common words that are capitalized (going for names here).
print([word for word in most_comm if len(word[0]) > 0 and word[0][0].isupper()])

# Find specific words
if len(sys.argv) > 1:
    word = sys.argv[1]
    if word in wordcount.keys():
        occurence = wordcount[word]
        if word.lower() in wordcount.keys():
            occurence += wordcount[word.lower()]
        print('The word: [', word, '] appeared: ', occurence, ' times')

File: journal_analysis.cpp
#include <chrono>
#include <experimental/filesystem>
#include <fstream>
#include <iostream>
#include <regex>
#include <unordered_map>
namespace fs = std::experimental::filesystem;
using namespace std::chrono;

bool isTextFile(const std::string& s, const std::string& ending) {
    if (s.length() >= ending.length()) {
        return (0 == s.compare(s.length() - ending.length(), ending.length(), ending));
    } else {
        return false;
    }
}

void traverse_files(
    const std::string& path,
    const std::string& file_extension,
    const std::regex& words_regex,
    std::unordered_map<std::string, int>* wordcount)
{
    std::string words;
    std::ifstream file_reader;
    int count = 0;
    int frequency = 100;

    // For each file / directory in path
    for (const auto& entry : fs::recursive_directory_iterator(path)) {
        std::string file = static_cast<std::string>(entry.path());
        // if the file is a txt file
        if (isTextFile(file, file_extension)) {
            file_reader.open(file);
            if (file_reader.is_open()) {
                while (getline(file_reader, words)) {
                    auto words_begin = std::sregex_iterator(
                        words.begin(),
                        words.end(),
                        words_regex);
                    auto words_end = std::sregex_iterator();
                    for (std::sregex_iterator i = words_begin; i != words_end; ++i)
                    {
                        std::unordered_map<std::string, int>& counts = *wordcount;
                        ++counts[(*i).str()];
                    }
                }
                file_reader.close();
            }
            count++;
        }
    }

    std::cout << "Number of entries: " << count << std::endl;

    // Sorts most frequent words from unordered map to be displayed
    count = 0;
    std::map<int, std::string> frequentWords;
    for (const auto& n : *wordcount) {
        count += n.second;
        if (n.second > frequency && isupper(n.first[0])) {
            frequentWords[n.second] = n.first;
        }
    }
    for (auto n = frequentWords.rbegin(); n != frequentWords.rend(); ++n) {
        std::cout << n->second << " : " << n->first << ", ";
    }
    std::cout << std::endl;
    std::cout << "Total words: " << count << std::endl;
    std::cout << "Unique words: " << wordcount->size() << std::endl;
}

void find_specific_words(
    int argc,
    char** argv,
    std::unordered_map<std::string, int> wordcount)
{
    int occurrence = 0;
    std::string word = argv[1];
    if (wordcount.count(word) > 0) {
        occurrence = wordcount[word];
    }
    std::transform(word.begin(), word.end(), word.begin(),
        [](unsigned char c){ return std::tolower(c); });
    if (wordcount.count(word) > 0) {
        occurrence += wordcount[word];
    }
    std::cout << "The word: " << argv[1] << " appeared: "
        << occurrence << " times" << std::endl;
}

int main(int argc, char** argv) {
    auto start = high_resolution_clock::now();

    const std::string path = "/mnt/c/Eric/Journal";
    const std::string file_extension = ".txt";
    const std::regex words_regex("[^\\s.,;!?()]+");
    std::unordered_map<std::string, int> wordcount;

    traverse_files(path, file_extension, words_regex, &wordcount);
    if (argc > 1) {
        find_specific_words(argc, argv, wordcount);
    }

    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start).count() / 1000000.0;
    std::cout << "Took: " << duration << " seconds" << std::endl;
    return 0;
}
